WalletWasabi/Helpers/PowerSaving/LinuxInhibitorTask.cs

```diff
@@ -0,0 +1,310 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Logging;
+using WalletWasabi.Microservices;
+
+namespace WalletWasabi.Helpers.PowerSaving
+{
+	/// <summary><c>systemd-inhibitor</c> API wrapper.</summary>
+	/// <remarks>Only works on Linux machines that use systemd.</remarks>
+	/// <seealso href="https://www.freedesktop.org/wiki/Software/systemd/inhibit/"/>
+	public class LinuxInhibitorTask : IPowerSavingInhibitorTask
+	{
+		/// <remarks>Guarded by <see cref="StateLock"/>.</remarks>
+		private bool _isDone;
+
+		/// <remarks>Use the constructor only in tests.</remarks>
+		internal LinuxInhibitorTask(InhibitWhat what, TimeSpan period, string reason, ProcessAsync process)
+		{
+			What = what;
+			BasePeriod = period;
+			Reason = reason;
+			Process = process;
+			Cts = new CancellationTokenSource(period);
+
+			Task = WaitAsync();
+		}
+
+		/// <summary>Linux GUI environments.</summary>
+		public enum GraphicalEnvironment
+		{
+			Gnome,
+			Mate,
+			Other,
+		}
+
+		[Flags]
+		public enum InhibitWhat
+		{
+			/// <summary>
+			/// Inhibits that the system goes into idle mode, possibly resulting in automatic system
+			/// suspend or shutdown depending on configuration.
+			/// </summary>
+			Idle = 1,
+
+			/// <summary>Inhibits system suspend and hibernation requested by (unprivileged) users.</summary>
+			Sleep = 2,
+
+			/// <summary>Inhibits high-level system power-off and reboot requested by (unprivileged) users.</summary>
+			Shutdown = 4,
+
+			All = Idle | Sleep | Shutdown
+		}
+
+		/// <remarks>Guards <see cref="_isDone"/>.</remarks>
+		private object StateLock { get; } = new();
+
+		public InhibitWhat What { get; }
+
+		/// <remarks>It holds: inhibitorEndTime = now + BasePeriod + ProlongInterval.</remarks>
+		public TimeSpan BasePeriod { get; }
+
+		/// <summary>Reason why the power saving is inhibited.</summary>
+		public string Reason { get; }
+		private ProcessAsync Process { get; }
+		private CancellationTokenSource Cts { get; }
+		private TaskCompletionSource StoppedTcs { get; } = new();
+		private Task Task { get; }
```

Comment:
This is unused.

Reply:
Yes, I was somewhat afraid that the task will be garbage collected before it ends. But in the hindsight it seems like a nonsense for several reasons, so I'll remove it.